File: src/main/java/com/machfour/macros/cli/Meals.java
package com.machfour.macros.cli;

import com.machfour.macros.linux.Config;
import com.machfour.macros.linux.LinuxDatabase;
import com.machfour.macros.objects.Meal;
import com.machfour.macros.storage.MacrosDataSource;
import com.machfour.macros.storage.MacrosDatabase;
import com.machfour.macros.util.DateStamp;

import java.io.PrintStream;
import java.sql.SQLException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Map;

import static com.machfour.macros.cli.CliMain.OUT;
import static com.machfour.macros.cli.CliMain.PROGNAME;

public class Meals extends CommandImpl {

    private static final String NAME = "meals";
    @Override
    public String name() {
        return NAME;
    }
    @Override
    public void printHelp(PrintStream out) {
        OUT.printf("Usage: %s %s [day]\n", PROGNAME, NAME);
    }
    @Override
    public void doAction(List<String> args) {
        if (args.contains("--help")) {
            printHelp(OUT);
            return;
        }

        // cases: day not specified vs day specified
        MacrosDatabase db = LinuxDatabase.getInstance(Config.DB_LOCATION);
        ArgParsing.Result dateArgument = ArgParsing.findArgument(args, 1);
        DateStamp d = null;
        switch (dateArgument.status()) {
            case NOT_FOUND:
                d = DateStamp.forCurrentDate();
                break;
            case OPT_ARG_MISSING:
                OUT.println("-d option requires a day specified");
                break;
            case ARG_FOUND:
                d = ArgParsing.dayStringParse(dateArgument.argument());
                break;
            default:
                OUT.println("Invalid date format: '" + dateArgument.argument() + "'.");
                break;
        }
        if (d != null) {
            printMealList(db, d);
        }

        //Meal toEdit = mealSpec.processedObject();
    }

    private void printMealList(MacrosDataSource db, DateStamp d) {
        try {
            Map<String, Meal> meals = db.getMealsForDay(d);
            if (meals.isEmpty()) {
                OUT.println("No meals recorded on " + CliUtils.prettyDay(d));
            } else {
                OUT.println("Meals recorded on " + CliUtils.prettyDay(d) + ":");
                OUT.println();
                OUT.println(String.format("%-16s %-16s", "Name", "Last Modified"));
                OUT.println("=================================");
                SimpleDateFormat dateFormat = new SimpleDateFormat();
                for (Meal m : meals.values()) {
                    Date mealDate = new Date(m.modifyTime()*1000);
                    OUT.println(String.format("%-16s %16s", m.getName(), dateFormat.format(mealDate)));
                }
            }
        } catch (SQLException e) {
            OUT.println("SQL Exception: " + e.getMessage());
        }

    }
}

File: macros-cli/src/main/kotlin/com/machfour/macros/cli/modes/Meals.kt
package com.machfour.macros.cli.modes

import com.machfour.datestamp.DateStamp
import com.machfour.datestamp.currentDateStamp
import com.machfour.macros.cli.CliConfig
import com.machfour.macros.cli.CommandImpl
import com.machfour.macros.cli.utils.ArgParsingResult
import com.machfour.macros.cli.utils.dayStringParse
import com.machfour.macros.cli.utils.findArgument
import com.machfour.macros.cli.utils.printlnErr
import com.machfour.macros.jvm.modifyInstant
import com.machfour.macros.queries.getMealsForDay
import com.machfour.macros.sql.SqlDatabase
import com.machfour.macros.sql.SqlException
import com.machfour.macros.util.fmt
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle

class Meals(config: CliConfig) : CommandImpl(config) {
    override val name = "meals"
    override val usage = "Usage: ${config.programName} $name [day]"

    override fun doAction(args: List<String>): Int {
        if (args.contains("--help")) {
            printHelp()
            return 2
        }

        // cases: day not specified vs day specified
        val ds = config.database
        val date = when (val dateArg = findArgument(args, 1)) {
            is ArgParsingResult.KeyValFound -> {
                dayStringParse(dateArg.argument) ?: run {
                    printlnErr("Invalid date format: '${dateArg.argument}'.")
                    return 1
                }
            }
            is ArgParsingResult.ValNotFound -> {
                printlnErr("-d option requires a day specified")
                return 1
            }
            else -> currentDateStamp()
        }
        return printMealList(ds, date)
        //Meal toEdit = mealSpec.processedObject();
    }

    private fun printMealList(db: SqlDatabase, d: DateStamp): Int {
        try {
            val meals = getMealsForDay(db, d)
            if (meals.isEmpty()) {
                println("No meals recorded on " + d.prettyPrint())
            } else {
                println("Meals recorded on " + d.prettyPrint() + ":")
                println()
                println("Name".fmt(16, true) + " " + "Last Modified".fmt(16, true))
                println("=================================")
                val dateFormat = DateTimeFormatter.ofLocalizedTime(FormatStyle.LONG)
                for (m in meals.values) {
                    val mealDate = m.modifyInstant.atZone(ZoneId.systemDefault())
                    println(m.name.fmt(16, true) + " " + dateFormat.format(mealDate).fmt(16))
                }
            }
        } catch (e: SqlException) {
            printlnErr("SQL Exception: ${e.message}")
            return 1
        }
        return 0
    }

}